### ISSUE TITLE
Uthernet I or II: Contiki's web-browser reports timeout or connection reset after Apple II power-cycle

### ISSUE THREAD

**Issue by tomcw** [AppleWin/AppleWin#1058]
Repro with Uthernet I:
- AppleWin 1.30.8.0 (dev build with Uthernet II card support) / Npcap 1.60 / Win7-64
- Boot Contiki 2.7
  - ETHCONFI: Select Uthernet I (slot 3)
  - IPCONFIG: Request IP address (and save)
- Web browse to: INFO.CERN.CH... which works OK
- F2 (power-cycle), and skip the previous ETHCONFI / IPCONFIG, since it's saved to disk
- Web browse to: INFO.CERN.CH... sometimes get: "Connection timed out" or "Connection reset by peer"
  - retrying succeeds after a few attempts

Repro with Uthernet II:
- Basically same as above, instead just select: Uthernet II (slot 3) + re-request IP address

**NB. This isn't a regression, as I have reproduced on AppleWin 1.25.0.4 / WinPcap 4.1.3 / Win7-64 with a Uthernet I card.**

**Comment by audetto** [comment]
Nothing obvious comes to my mind.
Looking at the save states files, they look identical. I have the same issues with the PCAP backend but not with libslirp.
Can you save 2 save state files at the webbrowser (before you try to request the page).

And if you reload the state file from the working session, does it work?

**Comment by tomcw** [comment]
In `ResetMachineState()` (ie. the the Apple II power-cycle function), I added this bit of debug code after `GetCardMgr().Reset(true)`:
```
	SS_CARDTYPE slot3card = GetCardMgr().QuerySlot(SLOT3);
	GetCardMgr().Remove(SLOT3, false);
	GetCardMgr().Insert(SLOT3, slot3card, false);
	GetCardMgr().GetRef(SLOT3).InitializeIO(NULL);
```

This forces the NetworkBackend to be deleted, and then `InitializeIO()` will re-enumerate the NIC adaptors, and reopen our one (pcap functions: open_live, setnonblock datalink).

This still fails in the same way.

**Comment by tomcw** [comment]
>And if you reload the state file from the working session, does it work?

Reloading a save-state gives the same failure.
(And it's basically the same as my bit of debug code that I added to `ResetMachineState()`, above.)

**Comment by tomcw** [comment]
In `TfePcapLoadLibrary()`, I tried having it initially call `TfePcapFreeLibrary()`, so that it always frees, then reloads the Npcap DLL.

This still fails in the same way.

I will try on a newer Windows OS (eg. Win10).

**Comment by audetto** [comment]
Could it be that PCAP needs some type of warm up? and diving directly into a webbrowser is not good?

**Comment by audetto** [comment]
I am not too sure what to say. Now it works. But I run inside a Windows Virtual Machine with QEMU...

**Comment by tomcw** [comment]
Another "data-point": I can reproduce the same issue by just doing CTRL+RESET (instead of a full Apple II power-cycle).
Again for both U1 and U2 cards.

NB. For U1, the `Reset(powerCycle=false)` does nothing. Only the `powerCycle=true` case will call `Init()`.
(For U2, `Reset(powerCycle=false)` does init a fair bit of W5100 h/w state.)

Since for U1, `Reset(powerCycle=false)` does nothing, then maybe it is an issue with Contiki itself.

**Comment by audetto** [comment]
If you could try U2, uncomment this block : https://github.com/AppleWin/AppleWin/blob/169f0b29e8e7b45757866b7cb0e815bc1aadf9dc/source/Uthernet2.cpp#L76-L79

And just send me the log of the 2nd run, after the reboot, when you go straight to the web-browser and possibly the pcap file from wireshark (ideally with little extra background network activity...)

**Comment by tomcw** [comment]
>I will try on a newer Windows OS (eg. Win10).

AppleWin 1.30.8.0 (HEAD dev build with Uthernet II card support) / Npcap 1.60 / **Win10**
(using USB3-to-ethernet adaptor)

U1 behaves as on Win7.
U2 sometimes works after a reset/power-cycle.

In both cases, if (after reset/power-cycle) I _first_ do an IPCONFIG ("Request IP address", but don't save it), then it seems OK in nearly all cases. IE, maybe this "Request IP address" puts the  emulated U1/U2 card into a "good" state.

I'll try the extra logging & pcap another day.

**Comment by tomcw** [comment]
Can I just use the aw.zip from this other issue: https://github.com/AppleWin/AppleWin/issues/1046#issuecomment-1066843270 ?

**Comment by audetto** [comment]
Yes

**Comment by tomcw** [comment]
OK, I used the filter "ip.src == 192.168.0.35 || ip.dst == 192.168.0.35", and exported just these packets to the pcap file:

[AppleWin-trace1_src-dst-ip-192_168_0_35-filtered.pcap.gz](https://github.com/AppleWin/AppleWin/files/8248841/AppleWin-trace1_src-dst-ip-192_168_0_35-filtered.pcap.gz)

What I did:
- WEBBROWS: INFO.CERN.CH  (this worked fine)
- CTRL+RESET
- WEBBROWS: INFO.CERN.CH  (this timed-out after 1 minute, with Contiki error: "Connection reset by peer")
- Immediately retry, and it succeeds

**Comment by tomcw** [comment]
And here's a similar filtered trace for Uthernet2:

[AppleWin-trace2_src-dst-ip-192_168_0_36-filtered.pcap.gz](https://github.com/AppleWin/AppleWin/files/8248920/AppleWin-trace2_src-dst-ip-192_168_0_36-filtered.pcap.gz)

What I did:

- Using Uthernet2, and IPCONFIG assigns ip=192.168.0.36
- WEBBROWS: INFO.CERN.CH (this worked fine)
- CTRL+RESET
- WEBBROWS: INFO.CERN.CH (this immediately reported Contiki error: "Connection reset by peer")
- Immediately retry, and it succeeds

**Comment by audetto** [comment]
I am no expert of TCP handshake.
But it should start with a `SYN`.

Pick trace1.

After then DNS request, packet 5 is the first successful connection starting with a `SYN`.

Then, at 16sec you must have done a restart and you see anew DNS query, but the following packet 30 is not `SYN` according to Wireshark, but some retransmission.

Only at sec 83 we see a genuine `SYN` (packet 52) which seems to start a good connection.

Similar pattern in trace 2, packet 5, 34 & 41

I think we need someone form contiki to comment.

**Comment by audetto** [comment]
https://www.techopedia.com/definition/23554/initial-sequence-numbers-isn

**Comment by audetto** [comment]
I've done the same with `ip65.dsk` from https://github.com/cc65/ip65 and the first `SYN` is a *random* number

3661365248
3877765120 (reset)
3791519744 (rerun)
1168637952 (reset)
3483238400 (reset)
....

**Comment by audetto** [comment]
I am trying a fix. What is the best way to get some entropy?
A good address to read the floating bus?

**Comment by audetto** [comment]
It is indeed a weak source, but it does something

[contiki-apple2.random.zip](https://github.com/AppleWin/AppleWin/files/8262437/contiki-apple2.random.zip)

**Comment by tomcw** [comment]
Uthernet1, using your "contiki-apple2.po".
I used the filter "ip.src == 192.168.0.35 || ip.dst == 192.168.0.35", and exported just these packets to the pcap file:

[AppleWin-trace4-u1-timeout-src-dst-192_168_0_35-filtered.pcap.gz](https://github.com/AppleWin/AppleWin/files/8274813/AppleWin-trace4-u1-timeout-src-dst-192_168_0_35-filtered.pcap.gz)

- Pre-setup the disk as: ETHCONFI:Uthernet1, and IPCONFIG assigns ip=192.168.0.35

What I did for the packet capture: (same as before)

- WEBBROWS: INFO.CERN.CH (this worked fine)
- CTRL+RESET
- WEBBROWS: INFO.CERN.CH (this timed-out after 1 minute, with Contiki error: "Connection reset by peer")
- Immediately retry, and it succeeds

**Comment by audetto** [comment]
I think it works, but the bus entropy is really small.
In your case, it has read `A0` 4 times in a row in both cases, there is little to do.

Look at the raw sequence number in packet 5 and 34, it is 2694881440 (`A0 A0 A0 A0`).

Oliver suggested to use cc65's `_randomize` function which reads `4E` and `4F` but it does not seem to be random at all.

Unless someone finds a real source of entropy, I don't know what else I can add.

**Comment by tomcw** [comment]
You could fill TEXT page 2 ($800-$BFF) with random data, then briefly change the video mode to "40-COL, TEXT, PAGE2" when generating this 4-byte sequence number.

btw, `4E` and `4F` only increment (as a 16-bit word) when going via the standard ROM keyboard handler, ie. the time entropy between key-presses. If the code doesn't call the ROM handler, then `4E` and `4F` don't change.

**Comment by audetto** [comment]
This is my final effort:

Use the same contiki which reads `C020` together with this

[aw.zip](https://github.com/AppleWin/AppleWin/files/8303869/aw.zip)

where I changed

```
inline DWORD getRandomTime();
static BYTE __stdcall IORead_C02x(WORD pc, WORD addr, BYTE bWrite, BYTE d, ULONG nExecutedCycles)
{
	if (!bWrite && addr == 0xC020)
	{
		DWORD random = getRandomTime();
		return (BYTE)random;
	}
	else
	{
		return IO_Null(pc, addr, bWrite, d, nExecutedCycles);
	}
}
```

the real fix is in contiki, if this is actually worth fixing.

**Comment by tomcw** [comment]
This new build (`aw.zip`) doesn't create an AppleWin.pcap file.

**Comment by audetto** [comment]
Here it is.

[aw.zip](https://github.com/AppleWin/AppleWin/files/8309326/aw.zip)

**Comment by tomcw** [comment]
I did 2 AppleWin pcap's with a U1 card, just directly booting Contiki each time, and then immediately web-browsing to INFO.CERN.CH, which always works without any Contiki errors.

1. Using _regular_ Contiki: Web Browser to INFO.CERN.CH
2. Using your [contiki-apple2.random.zip](https://github.com/AppleWin/AppleWin/files/8262437/contiki-apple2.random.zip), together with an AppleWin that returns `rand()` when $C020 is read.

From these 2 WireShark screenshots, you can see that they basically give the same results:
(I've highlighted the Sequence Number)
![test1](https://user-images.githubusercontent.com/6696896/159122881-220fa890-cb44-4421-a116-47859f791663.png)
![test2](https://user-images.githubusercontent.com/6696896/159122885-44cf9673-b4b6-44bd-898d-6cac2a34289b.png)

All those black lines saying "TCP Dup ACK" don't give me much confidence, ie. a "black row" looks like an error, but it's just a note... `[Expert Info (Note/Sequence): Duplicate ACK (#1)]`.

I plan to try to repro this (and pcap) on real U2 hardware.

**Comment by audetto** [comment]
I don't understand you. What has changed?

I would only look at the very first connection, the one after the DNS request.
With my contiki + applewin it should always (99.99%) of the times have different `Sequence Number (raw)`,make sure you look at **raw**.

Wireshark still objects that we are reusing a port, I don't know, maybe the port should change too for a well behaved TCP stack. I guess it is possible that a server will mistake this for bad traffic if it arrives from a port which has just been used. But it seems to work.

All the other black lines might be caused by the extremely slowness of AW or the fact that we miss a lot of replies and the server needs to retransmit.

What are your concerns exactly?

**Comment by audetto** [comment]
[contiki-apple2.random.zip](https://github.com/AppleWin/AppleWin/files/8309408/contiki-apple2.random.zip)

Try this, it randomises the source port too.

**Comment by tomcw** [comment]
Trying with real U2 hardware...

1. Web Browse INFO.CERN.CH  - 1st, 2nd, 3rd, ... are all OK!
2. CTRL+RESET, then 45s to reload WEBBROWS
3. Web Browse INFO.CERN.CH  - 1st is OK, but 2nd attempt "Connection reset by peer", then 3rd is OK
4. CTRL+RESET, then 45s to reload WEBBROWS
5. Web Browse INFO.CERN.CH  - 1st, 2nd, 3rd,... 9th are all "Connection reset by peer", then it's OK
6. CTRL+RESET, then 45s to reload WEBBROWS
7. Web Browse INFO.CERN.CH  - 1st, 2nd, 3rd, ... are all OK!
8. CTRL+RESET, then 45s to reload WEBBROWS
9. Web Browse INFO.CERN.CH  - 1st, 2nd, 3rd,... 5th are all "Connection reset by peer", then it's OK

>What are your concerns exactly?

Above I didn't see why the Sequence Number had to be random.
I thought: surely 0x00000001 (Test1.pcap) is as good as 0xBA13ACA9 (Test2.pcap)?
Anyway, I understand better - see [1] below.

But perhaps the web server keeps a record of previous sequence numbers, and that's why I get the Contiki error: "Connection reset by peer". Obviously the web server can't keep a record of all earlier sequence numbers, so maybe after some time (eg. 30s) it forgets these older sequence numbers, freeing up space in its cache.

This would explain why sometimes it works OK after a CTRL+RESET and other times it gets "Connection reset by peer".

NB. CTRL+RESET and reloading WEBBROWS will reset the sequence number back to 0x00000001.
[1] So for this reason (and the web server having cache of previous sequence numbers) means a random seq number must be used.

Anyway, clearly Contiki + real U2 card are behaving similar (the same?) to the emulator. And as you say, the fix needs to be done in Contiki itself.

**Comment by tomcw** [comment]
>Try this, it randomises the source port too.

Same result as above, ie. for the initial Web Browse to INFO.CERN.CH I get the same pattern of black lines saying "TCP Dup ACK".

>All the other black lines might be caused by the extremely slowness of AW

I tried running AppleWin at full-speed, but again got the same pattern of black lines saying "TCP Dup ACK".

**Comment by tomcw** [comment]
I think I can probably close this issue... in summary it's a Contiki issue.
But I would like to get a packet capture from a real Apple //e with a real U2 card.

**Comment by audetto** [comment]
Definitely, TCP needs to handle lost and duplicate packets so it needs to know what belongs to which connection and when a new one starts.

Probably a new local port is enough, and would let the previous one time out.

It seems that a brand new sequence number does it too.

If you can capture the traffic of the real U2 card, it would be interesting to see what the hardware TCP flow control does: 

[ip65.zip](https://github.com/AppleWin/AppleWin/files/8309570/ip65.zip)

Try to call WGET65 and see what sequence number and port the card uses. Maybe it contains some sort of random number generator.

**Comment by audetto** [comment]
See how ip65 does it, for software tcp:

https://github.com/cc65/ip65/blob/fba9b90cde1ec7374c505ce2ee372798bd8173ab/ip65/tcp.s#L167-L170

and

https://github.com/cc65/ip65/blob/fba9b90cde1ec7374c505ce2ee372798bd8173ab/ip65/ip65.s#L46
